### DISCUSSION TITLE
Add a way of protecting databases

### DISCUSSION THREAD

**Discussion by peskybogle** [NativePHP/laravel#385] [Ideas]
It would be good if, as well as protecting source code, we could protect our databases from easy user access. I collect quite a lot of data to run my app, I'd like to make it as hard as possible for someone to just take it all.

SQLCipher came up as a SQLite/PHP friendly option from a quick search, but it requires PHP extensions that aren't included with NativePHP by default. I'm not sure if it's possible to make it work with NativePHP, but it would be good if someone far more competent than me could look.

**Comment by simonhamp** [comment]
As we discussed elsewhere, there are a couple of challenges here:

1. Custom extensions for NativePHP's PHP binaries isn't yet supported, though I do believe it is technically possible to _dynamically_ load extensions even into the _statically-compiled_ binaries.
2. You'd need to compile the extension for each platform you wish to support and dynamically [wire up the php.ini](https://nativephp.com/docs/1/getting-started/configuration#customize-phpini) in order for the relevant platform-specific binary of PHP to load the correct version of the extension.
3. Any encryption on the client is going to require the decryption key to live on the client too, which means the protection is only "skin deep" for anyone who's really intent on decrypting the data.
